dmd.root.array: add any and all predicates to dmd.root.array

gorsing opened the pull request:
Description:
Adds high-performance, inlined any and all predicates to Array in dmd.root.array. These templates use alias predicates to ensure zero-overhead execution, facilitating the refactoring of manual for loops in semantic analysis stages.

dkorpel reviewed (changes requested):
Thanks for the examples. Unfortunately I don't think they demonstrate benefits. It would be helpful if it collapsed complex loop logic into something simpler, but replacing a foreach loop body with a function literal passed to a call to `all` or `any` only makes the code harder to read in my opinion, as it's now harder to recognize that a loop is being performed.

gorsing commented:
I have taken out some parts of MR as separate problems
https://github.com/dlang/dmd/pull/22300
https://github.com/dlang/dmd/pull/22301

dkorpel commented:
> I have taken out some parts of MR as separate problems

Thanks for doing that!

dkorpel commented:
> I have updated our base Array!T container with several high-performance methods O(N).

Do you have benchmarks showing how much this improves compilation times? Note that the dmd.root.array module is not a complete general purpose array implementation, it is only used for dmd to make bootstrapping easier. Hence additions to that module need to be motivated through dmd's use cases.

Commit:
refactoring add any && all

gorsing commented:
Hi @dkorpel 
I've highlighted the changes to the old code in https://github.com/dlang/dmd/pull/22327 to separate the addition of new methods from fixes or improvements.

gorsing commented:
> > I have updated our base Array!T container with several high-performance methods O(N).
> 
> Do you have benchmarks showing how much this improves compilation times? Note that the dmd.root.array module is not a complete general purpose array implementation, it is only used for dmd to make bootstrapping easier. Hence additions to that module need to be motivated through dmd's use cases.

I am currently working on a profiler report to demonstrate the specific performance gains.

Regarding total compilation time: measuring a macro-level improvement right now is non-trivial. Since I extracted the refactoring of semantic3.d into a separate PR to keep this contribution minimal, these new methods are not yet being utilized in the compiler's hot paths. Consequently, any immediate gains are currently lost in the noise of a full build.

However, the report I am preparing compares the current O(N²) patterns often found in DMD (manual loops with remove) against the O(N) removeIf implementation. My goal is to show that while dmd.root.array is indeed specialized, adding these algorithmic primitives provides the necessary infrastructure to optimize.

I will post the benchmark results shortly.

dkorpel commented:
> Consequently, any immediate gains are currently lost in the noise of a full build.

It is indeed hard, but important to measure because expectations and reality can wildly differ when trying to optimize something. 

> I will post the benchmark results shortly.

Great! If you need any help benchmarking let me know.
